### PR TITLE
sql runner

### PR DESCRIPTION
**Pull request by pyr**
A minimalist job queue worker for SQL. This assumes multiple worker off of a main job queue. In the following code, each worker is assumed to be able to process both actions and workflow evaluation. The processing is based off of two tables: `task` and `workflow_run`. `workflow_run` holds information about a full workflow run and its current status, while `task` is used for both calls to `job/restart` and action dispatches.

@mping-exo @tdantas note that this is early work after our discussions in Prague. I wanted to exchange on the idea and to see whether this helps slowly transition over the ablauf2 repository. I built this on the way back and it works for the tests scenarios I have locally.

You can check the comment block for how to use it to start a processing thread. It needs improving especially on the runner function for actions but can maybe used as an inspiration.

A few things:

- A single table for all tasks, no task type per leaf only to separate whether the task is a workflow restart action or a leaf execution. Maybe later on that type can be used to have workers dedicated to certain leaf types, but for now it feels premature
- The `workflow_run` acts a bit the same than the cassandra jobstore does (pushing changes to cassandra could be done via kafka connect to centralize all jobs).
- This is obviously still missing some things we talked about and are being worked on in ablauf2:
  - [x] tests
  - [ ] efficient serialization
  - [ ] state simplification
  - [ ] state transition checks
  
The flipside is that this can help run performance assessments really quickly.

### CODE REVIEW

File: schema.sql [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    updated_at timestamp     NOT NULL DEFAULT now() ON UPDATE NOW(),
+    PRIMARY KEY(id)
+);
+
+CREATE TABLE IF NOT EXISTS task (
+    id         int         AUTO_INCREMENT,
+    type       enum('workflow', 'action') NOT NULL,
+    wid        int,
+    payload    text,
+    process_at timestamp   NOT NULL DEFAULT now(),
```
[the remark this answers is not in the document]
Comment (pyr) [reply schema.sql:20]: Yes, to be honest I have given no thought to indices yet I wanted to make sure the SQL approach was solid. I will add stories to schedule work on performance relevant tasks.

**Comment by mping-exo** [comment]
@pyr @tdantas some notes:

- tests are passing with both manifold and sql runner

- created a macro to run parameterized tests: `(deftestp expected-failure-test [arunner [runner sql-runner]]`
	- expands to something like
```
(doseq [arunner [runner sql-runner]]
  (testing (str "arunner:"arunner) ....))
```

- deal with `runtime` on job restart
  -  I decided to ditch saving the jobid on the runtime to ensure tests pass
  - the `exec/context` wasnt being propagated, same for runtime
- serialize `(with-augment [identity :out])`
	-  experimented with using vars `#identity` and custom edn readers, opted by using `::ast/identity` as it seems our codebase only uses `identity` as augment src, so we  would need to migrate those jobs to use `[::ast/identity ...]` instead

We should need to deal with:
- failure of the main worker `loop`: it will hang the worker thread and no job will finish
- some kind of logging to ensure minimal visibility in case everything fails

File: schema.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (pyr) [reply schema.sql]: The indices have been added, thanks!

**Review by mping-exo — APPROVED** [review]
:shipit:

**Pull request merged**